src/nvim/eval.c

```diff
@@ -8018,26 +8019,36 @@ static void f_bufnr(typval_T *argvars, typval_T *rettv)
     rettv->vval.v_number = -1;
 }
 
-/*
- * "bufwinnr(nr)" function
- */
-static void f_bufwinnr(typval_T *argvars, typval_T *rettv)
+static void buf_win_common(typval_T *argvars, typval_T *rettv, bool get_nr)
 {
-  (void)get_tv_number(&argvars[0]);         /* issue errmsg if type error */
-  ++emsg_off;
+  (void)get_tv_number(&argvars[0]);  // issue errmsg if type error
```

Comment:
The error message is printed in `get_tv_number_chk` if the types do not match. The documentation say that `v:true` is converted to `"true"` when used as a String and to 1 when used as a Number, so it makes sense that `:echo bufwinid(v:true)` doesn't give an error message, since it is converted to 1.

But your point still stands, we can just return -1 for other cases where the argument is not valid. This can be done by calling `get_tv_number_chk` directly and checking whether an error occurred.